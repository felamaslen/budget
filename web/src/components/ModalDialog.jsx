/**
 * Displays a modal dialog for editing / adding content
 */

import React from 'react';
import PropTypes from 'prop-types';
import PureControllerView from './PureControllerView';
import { List as list } from 'immutable';
import classNames from 'classnames';

import getFormField from './FormField';

import { aMobileDialogClosed } from '../actions/FormActions';

export class ModalDialog extends PureControllerView {
    renderTitle() {
        if (this.props.id) {
            return `Editing id#${this.props.id}`;
        }

        return 'Add item';
    }
    renderFields() {
        return this.props.fields
            .map((field, fieldKey) => {
                const item = getFormField(this.props.dispatcher, {
                    fieldKey,
                    item: field.get('item'),
                    value: field.get('value')
                });

                const invalid = this.props.invalidKeys.includes(fieldKey);

                const className = classNames({
                    'form-row': true,
                    invalid
                });

                return <li key={fieldKey} className={className}>
                    <span className="form-label">{field.get('item')}</span>
                    {item}
                </li>;
            });
    }
    renderButtons() {
        const onCancel = () => this.props.dispatcher.dispatch(aMobileDialogClosed(null));
        const onSubmit = () => this.props.dispatcher.dispatch(aMobileDialogClosed(
            this.props.pageIndex
        ));

        return <div className="buttons">
            <button type="button" className="button-cancel"
                onClick={onCancel}>nope.avi</button>
            <button type="button" className="button-submit"
                onClick={onSubmit}>Do it.</button>
        </div>;
    }
    render() {
        const className = `modal-dialog-outer ${this.props.type}`;
        const title = this.renderTitle();
        const fields = this.renderFields();
        const buttons = this.renderButtons();

        return <div className={className}>
            <div className="dialog">
                <span className="title">{title}</span>
                <ul className="form-list">
                    {fields}
                </ul>
                {buttons}
            </div>
        </div>;
    }
}

ModalDialog.propTypes = {
    type: PropTypes.string,
    pageIndex: PropTypes.number,
    row: PropTypes.number,
    col: PropTypes.number,
    id: PropTypes.number,
    fields: PropTypes.instanceOf(list),
    invalidKeys: PropTypes.instanceOf(list)
}

