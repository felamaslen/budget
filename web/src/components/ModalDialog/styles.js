import styled from 'styled-components';

export const ModalDialog = styled.div``;
