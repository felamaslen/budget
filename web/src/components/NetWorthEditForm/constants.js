export const STEP_DATE = 'STEP_DATE';
export const STEP_CURRENCIES = 'STEP_CURRENCIES';
export const STEP_VALUES = 'STEP_VALUES';
export const STEP_ASSETS = 'STEP_ASSETS';
export const STEP_LIABILITIES = 'STEP_LIABILITIES';

export const steps = [STEP_DATE, STEP_CURRENCIES, STEP_ASSETS, STEP_LIABILITIES];
