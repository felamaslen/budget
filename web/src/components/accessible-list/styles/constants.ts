export const fieldSizes: { [field: string]: number } = {
  date: 100,
  cost: 96,
  default: 180,
};

export const rowHeightMobile = 30;
export const rowHeightDesktop = 24;
