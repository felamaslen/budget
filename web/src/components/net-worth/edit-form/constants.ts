export const enum Step {
  Date,
  Currencies,
  Assets,
  Liabilities,
}

export const steps: Step[] = [Step.Date, Step.Currencies, Step.Assets, Step.Liabilities];
