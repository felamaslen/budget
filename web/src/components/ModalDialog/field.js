import React, { useCallback } from 'react';
import PropTypes from 'prop-types';

import FormFieldText from '~client/components/FormField';
import FormFieldDate from '~client/components/FormField/date';
import FormFieldCost from '~client/components/FormField/cost';
import FormFieldTransactions from '~client/components/FormField/transactions';
import * as Styled from './styles';

const FormFieldContainer = ({ children, item }) => (
    <Styled.FormRow item={item}>
        <Styled.FormLabel>{item}</Styled.FormLabel>
        {children}
    </Styled.FormRow>
);

FormFieldContainer.propTypes = {
    item: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
};

export default function ModalDialogField({ item, value, invalid, onChange }) {
    const onChangeCallback = useCallback(newValue => onChange(item, newValue), [onChange, item]);

    if (item === 'date') {
        return (
            <FormFieldContainer item={item}>
                <FormFieldDate invalid={invalid} value={value} onChange={onChangeCallback} />
            </FormFieldContainer>
        );
    }
    if (item === 'cost') {
        return (
            <FormFieldContainer item={item}>
                <FormFieldCost invalid={invalid} value={value} onChange={onChangeCallback} />
            </FormFieldContainer>
        );
    }
    if (item === 'transactions') {
        return (
            <Styled.FormRow item={item}>
                <Styled.FormRowInner item={item}>
                    <Styled.FormLabel item={item}>{item}</Styled.FormLabel>
                    <FormFieldTransactions
                        invalid={invalid}
                        value={value}
                        onChange={onChangeCallback}
                        active
                    />
                </Styled.FormRowInner>
            </Styled.FormRow>
        );
    }

    return (
        <FormFieldContainer item={item}>
            <FormFieldText invalid={invalid} value={value} onChange={onChangeCallback} />
        </FormFieldContainer>
    );
}

ModalDialogField.propTypes = {
    item: PropTypes.string.isRequired,
    value: PropTypes.any,
    invalid: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
};
