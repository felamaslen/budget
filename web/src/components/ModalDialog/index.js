import React, {
    useRef,
    useState,
    useReducer,
    useEffect,
    useCallback,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import compose from 'just-compose';
import { DateTime } from 'luxon';

import { replaceAtIndex } from '~client/modules/data';
import { validateField } from '~client/modules/validate';
import ModalDialogField from '~client/components/ModalDialog/field';

import * as Styled from './styles';
import './style.scss';

function getTitle({ type, id }) {
    if (type === 'edit') {
        return `Editing id#${id}`;
    }

    return 'Add item';
}

export const animationTime = 350;

const withDefaultDate = fields =>
    replaceAtIndex(
        fields,
        fields.findIndex(({ item, value }) => item === 'date' && !value),
        { item: 'date', value: DateTime.local() },
    );

const HIDDEN = 'HIDDEN';
const SHOWN = 'SHOWN';
const CHANGED_ID = 'CHANGED_ID';

const updatePersistentState = (state, { payload }) => ({
    ...state,
    title: getTitle(payload),
    canRemove: Boolean(payload.onRemove),
});

function reducer(state, action) {
    if (action.type === HIDDEN) {
        return {
            ...state,
            visible: false,
        };
    }
    if (action.type === SHOWN) {
        return {
            ...updatePersistentState(state, action),
            visible: true,
        };
    }
    if (action.type === CHANGED_ID) {
        return updatePersistentState(state, action);
    }

    return state;
}

export default function ModalDialog({
    active,
    loading,
    id,
    fields,
    type,
    onCancel,
    onSubmit,
    onRemove,
}) {
    const [state, dispatch] = useReducer(reducer, {
        title: getTitle({ type, id }),
        canRemove: Boolean(onRemove),
        visible: active,
    });

    const timer = useRef(null);
    const { title, canRemove, visible } = state;

    useEffect(() => {
        if (active && !visible) {
            dispatch({ type: SHOWN, payload: { type, id, onRemove } });
        } else if (!active && visible) {
            clearTimeout(timer.current);
            timer.current = setTimeout(
                () => dispatch({ type: HIDDEN }),
                animationTime,
            );
        } else {
            dispatch({ type: CHANGED_ID, payload: { type, id, onRemove } });
        }
    }, [type, id, onRemove, active, visible]);

    useEffect(() => () => clearTimeout(timer.current), []);

    const [tempFields, setTempFields] = useState();
    useEffect(() => {
        setTempFields(compose(withDefaultDate)(fields.slice()));
    }, [fields]);

    const [invalid, setInvalid] = useState({});

    const onChangeField = useCallback(
        (item, value) =>
            setTempFields(last =>
                replaceAtIndex(
                    last,
                    last.findIndex(({ item: thisItem }) => thisItem === item),
                    { item, value },
                ),
            ),
        [],
    );

    const onSubmitCallback = useCallback(() => {
        const nextInvalid = tempFields.reduce((last, { item, value }) => {
            try {
                validateField(item, value);

                return last;
            } catch (err) {
                return { ...last, [item]: true };
            }
        }, {});

        setInvalid(nextInvalid);

        if (!Object.keys(nextInvalid).length) {
            onSubmit(
                tempFields.reduce(
                    (last, { item, value }) => ({ ...last, [item]: value }),
                    { id },
                ),
            );
        }
    }, [onSubmit, tempFields, id]);

    useEffect(() => {
        if (!active && Object.keys(invalid).length) {
            setInvalid({});
        }
    }, [active, invalid]);

    if (!visible) {
        return null;
    }

    return (
        <Styled.ModalDialog className={classNames('modal-dialog', type)}>
            <div
                className={classNames('modal-dialog-inner', {
                    hidden: !active,
                    loading,
                })}
            >
                <span className="title">{title}</span>
                <ul className="form-list">
                    {fields.map(({ item, value }) => (
                        <ModalDialogField
                            key={item}
                            item={item}
                            value={value}
                            invalid={Boolean(invalid[item])}
                            onChange={onChangeField}
                        />
                    ))}
                </ul>
                <div className="buttons">
                    <button
                        type="button"
                        className="button-cancel"
                        disabled={loading}
                        onClick={onCancel}
                    >
                        {'nope.avi'}
                    </button>
                    <button
                        type="button"
                        className="button-submit"
                        disabled={loading}
                        onClick={onSubmitCallback}
                    >
                        {'Do it.'}
                    </button>
                    {canRemove && (
                        <button
                            type="button"
                            className="button-remove"
                            disabled={loading}
                            onClick={onRemove}
                        >
                            &minus;
                        </button>
                    )}
                </div>
            </div>
        </Styled.ModalDialog>
    );
}

ModalDialog.propTypes = {
    active: PropTypes.bool.isRequired,
    loading: PropTypes.bool,
    type: PropTypes.string,
    id: PropTypes.string,
    fields: PropTypes.arrayOf(
        PropTypes.shape({
            item: PropTypes.string.isRequired,
            value: PropTypes.any,
        }),
    ),
    onCancel: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
    onRemove: PropTypes.func,
};

ModalDialog.defaultProps = {
    active: true,
    loading: false,
    type: 'edit',
    id: null,
    fields: [],
    onRemove: null,
};
