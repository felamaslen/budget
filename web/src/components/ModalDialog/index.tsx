import React, { useRef, useState, useReducer, useEffect, useCallback } from 'react';
import { compose } from '@typed/compose';
import { DateTime } from 'luxon';
import { replaceAtIndex } from 'replace-array';

import { validateField } from '~client/modules/validate';
import { Button, ButtonSubmit, ButtonCancel } from '~client/styled/shared/button';
import ModalDialogField from './field';

import * as Styled from './styles';

type State = {
  visible: boolean;
  title: string;
  canRemove: boolean;
};

enum ActionType {
  Hidden,
  Shown,
  ChangedId,
}

type PersistentStatePayload = {
  type: Props['type'];
  id?: string;
  canRemove: boolean;
};

type ActionHidden = {
  type: ActionType.Hidden;
};

type ActionShown = {
  type: ActionType.Shown;
  payload: PersistentStatePayload;
};

type ActionChangedId = {
  type: ActionType.ChangedId;
  payload: PersistentStatePayload;
};

type Action = ActionHidden | ActionShown | ActionChangedId;

function getTitle({ type, id }: Pick<PersistentStatePayload, 'type' | 'id'>): string {
  if (type === 'edit') {
    return `Editing id#${id}`;
  }
  return 'Add item';
}

export const animationTime = 350;

type Field = {
  item: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value?: any;
};

const withDefaultDate = (fields: Field[]): Field[] =>
  replaceAtIndex(
    fields,
    fields.findIndex(({ item, value }) => item === 'date' && !value),
    {
      item: 'date',
      value: DateTime.local(),
    },
  );

const updatePersistentState = (
  state: State,
  { payload }: { payload: PersistentStatePayload },
): State => ({
  ...state,
  title: getTitle(payload),
  canRemove: payload.canRemove,
});

type Reducer = (state: State, action: Action) => State;

const reducer: Reducer = (state, action): State => {
  if (action.type === ActionType.Hidden) {
    return {
      ...state,
      visible: false,
    };
  }
  if (action.type === ActionType.Shown) {
    return {
      ...updatePersistentState(state, action),
      visible: true,
    };
  }
  if (action.type === ActionType.ChangedId) {
    return updatePersistentState(state, action);
  }

  return state;
};

export type Props = {
  active: boolean;
  loading?: boolean;
  type?: 'edit' | 'add';
  id?: string;
  fields?: Field[];
  onCancel: () => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onSubmit: (item: any) => void;
  onRemove?: () => void;
};

const ModalDialog: React.FC<Props> = ({
  active,
  loading = false,
  id,
  fields = [],
  type = 'edit',
  onCancel,
  onSubmit,
  onRemove,
}) => {
  const canRemove = !!onRemove;
  const [state, dispatch] = useReducer<Reducer>(reducer, {
    title: getTitle({ type, id }),
    canRemove,
    visible: active,
  });

  const timer = useRef<number>();
  const { title, visible } = state;

  useEffect(() => {
    if (active && !visible) {
      dispatch({ type: ActionType.Shown, payload: { type, id, canRemove } });
    } else if (!active && visible) {
      clearTimeout(timer.current);
      timer.current = setTimeout(() => dispatch({ type: ActionType.Hidden }), animationTime);
    } else {
      dispatch({ type: ActionType.ChangedId, payload: { type, id, canRemove } });
    }
  }, [type, id, canRemove, active, visible]);

  useEffect((): (() => void) => (): void => clearTimeout(timer.current), []);

  const [tempFields, setTempFields] = useState<Field[]>([]);
  useEffect(() => {
    setTempFields(compose(withDefaultDate)(fields.slice()));
  }, [fields]);

  const [invalid, setInvalid] = useState<Record<string, boolean>>({});

  const onChangeField = useCallback(
    (item, value) =>
      setTempFields(last =>
        replaceAtIndex(
          last,
          last.findIndex(({ item: thisItem }) => thisItem === item),
          {
            item,
            value,
          },
        ),
      ),
    [],
  );

  const onSubmitCallback = useCallback(() => {
    const nextInvalid = tempFields.reduce((last, { item, value }) => {
      try {
        validateField(item, value);

        return last;
      } catch (err) {
        return { ...last, [item]: true };
      }
    }, {});

    setInvalid(nextInvalid);

    if (!Object.keys(nextInvalid).length) {
      onSubmit(tempFields.reduce((last, { item, value }) => ({ ...last, [item]: value }), { id }));
    }
  }, [onSubmit, tempFields, id]);

  useEffect(() => {
    if (!active && Object.keys(invalid).length) {
      setInvalid({});
    }
  }, [active, invalid]);

  if (!visible) {
    return null;
  }

  return (
    <Styled.ModalDialog>
      <Styled.ModalInner active={active} isLoading={loading}>
        <Styled.Title>{title}</Styled.Title>
        <Styled.FormList data-testid="form-list">
          {fields.map(({ item, value }) => (
            <ModalDialogField
              key={item}
              item={item}
              value={value}
              invalid={Boolean(invalid[item])}
              onChange={onChangeField}
            />
          ))}
        </Styled.FormList>
        <Styled.Buttons>
          <ButtonCancel type="button" disabled={loading} onClick={onCancel}>
            {'nope.avi'}
          </ButtonCancel>
          <ButtonSubmit type="button" disabled={loading} onClick={onSubmitCallback}>
            {'Do it.'}
          </ButtonSubmit>
          {state.canRemove && (
            <Button type="button" disabled={loading} onClick={onRemove}>
              &minus;
            </Button>
          )}
        </Styled.Buttons>
      </Styled.ModalInner>
    </Styled.ModalDialog>
  );
};

export default ModalDialog;
