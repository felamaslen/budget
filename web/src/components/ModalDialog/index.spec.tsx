import sinon from 'sinon';
import { render, fireEvent, act } from '@testing-library/react';
import React from 'react';

import ModalDialog, { Props, animationTime } from '.';
import { CREATE_ID } from '~client/constants/data';

describe('<ModalDialog />', () => {
  const props: Props = {
    active: true,
    loading: false,
    id: 'some-id',
    fields: [
      { item: 'item', value: 'some item' },
      { item: 'cost', value: 342 },
    ],
    type: 'edit',
    onCancel: jest.fn(),
    onSubmit: jest.fn(),
  };

  it('should hide after a delay', () => {
    const clock = sinon.useFakeTimers();

    const { container } = render(<ModalDialog {...props} />);
    expect(container.childNodes).toHaveLength(1);

    act(() => {
      render(<ModalDialog {...props} active={false} />, { container });
    });
    expect(container.childNodes).toHaveLength(1);

    act(() => {
      clock.tick(animationTime - 1);
    });
    expect(container.childNodes).toHaveLength(1);

    act(() => {
      clock.tick(1);
    });
    expect(container.childNodes).toHaveLength(0);

    clock.restore();
  });

  it('should show from inactive', () => {
    const { container } = render(<ModalDialog {...props} active={false} />);
    expect(container.childNodes).toHaveLength(0);

    act(() => {
      render(<ModalDialog {...props} active />, { container });
    });
    expect(container.childNodes).toHaveLength(1);
  });

  it('should render a title', () => {
    const { getByText } = render(<ModalDialog {...props} />);
    expect(getByText('Editing id#some-id')).toBeInTheDocument();
  });

  describe('when adding', () => {
    const propsAdding: Props = { ...props, id: CREATE_ID, type: 'add' };

    it('should render an alternative title', () => {
      const { getByText } = render(<ModalDialog {...propsAdding} />);
      expect(getByText('Add item')).toBeInTheDocument();
    });
  });

  it('should render a form list', () => {
    const { getByTestId } = render(<ModalDialog {...props} />);
    const formList = getByTestId('form-list');

    expect(formList).toBeInTheDocument();
    expect(formList).toMatchSnapshot();
  });

  describe('cancel button', () => {
    it('should be rendered', () => {
      const { getByText } = render(<ModalDialog {...props} />);
      const cancelButton = getByText('nope.avi');
      expect(cancelButton).toBeInTheDocument();
    });

    it('should call the onCancel event when clicked', () => {
      const { getByText } = render(<ModalDialog {...props} />);
      const cancelButton = getByText('nope.avi');

      act(() => {
        fireEvent.click(cancelButton);
      });
      expect(props.onCancel).toHaveBeenCalledTimes(1);
    });
  });

  describe('submit button', () => {
    it('should be rendered', () => {
      const { getByText } = render(<ModalDialog {...props} />);
      const submitButton = getByText('Do it.');
      expect(submitButton).toBeInTheDocument();
    });

    it('should call the onSubmit event when clicked', () => {
      const { getByText } = render(<ModalDialog {...props} />);
      const submitButton = getByText('Do it.');

      act(() => {
        fireEvent.click(submitButton);
      });
      expect(props.onSubmit).toHaveBeenCalledTimes(1);
      expect(props.onSubmit).toHaveBeenCalledWith({
        id: 'some-id',
        item: 'some item',
        cost: 342,
      });
    });

    it('should submit an edited form', () => {
      const { getByText, getByDisplayValue } = render(<ModalDialog {...props} />);
      const submitButton = getByText('Do it.');

      const inputItem = getByDisplayValue('some item');
      const inputCost = getByDisplayValue('3.42');

      expect(inputItem).toBeInTheDocument();
      expect(inputCost).toBeInTheDocument();

      act(() => {
        fireEvent.change(inputItem, { target: { value: 'other item' } });
      });
      act(() => {
        fireEvent.blur(inputItem);
      });
      act(() => {
        fireEvent.change(inputCost, { target: { value: '1.08' } });
      });
      act(() => {
        fireEvent.blur(inputCost);
      });

      act(() => {
        fireEvent.click(submitButton);
      });
      expect(props.onSubmit).toHaveBeenCalledTimes(1);
      expect(props.onSubmit).toHaveBeenCalledWith({
        id: 'some-id',
        item: 'other item',
        cost: 108,
      });
    });

    it('should not submit invalid values', () => {
      const { getByText, getByDisplayValue } = render(<ModalDialog {...props} />);
      const submitButton = getByText('Do it.');

      const inputItem = getByDisplayValue('some item');
      const inputCost = getByDisplayValue('3.42');

      expect(inputItem).toBeInTheDocument();
      expect(inputCost).toBeInTheDocument();

      act(() => {
        fireEvent.change(inputItem, { target: { value: '' } });
      });
      act(() => {
        fireEvent.blur(inputItem);
      });

      act(() => {
        fireEvent.click(submitButton);
      });
      expect(props.onSubmit).not.toHaveBeenCalled();
    });
  });

  describe('remove button', () => {
    it('should not be rendered if onRemove is not passed', () => {
      const { queryByText } = render(<ModalDialog {...props} onRemove={undefined} />);
      const removeButton = queryByText('−');
      expect(removeButton).not.toBeInTheDocument();
    });

    it('should be rendered if onRemove is passed', () => {
      const onRemove = jest.fn();
      const { getByText } = render(<ModalDialog {...props} onRemove={onRemove} />);
      const removeButton = getByText('−');
      expect(removeButton).toBeInTheDocument();
    });

    it('should call the onRemove event when clicked', () => {
      const onRemove = jest.fn();
      const { getByText } = render(<ModalDialog {...props} onRemove={onRemove} />);
      const removeButton = getByText('−');

      act(() => {
        fireEvent.click(removeButton);
      });
      expect(onRemove).toHaveBeenCalledTimes(1);
    });
  });

  describe('while loading', () => {
    const propsLoading = { ...props, onRemove: jest.fn(), loading: true };

    it('should disable the buttons', () => {
      const { getByText } = render(<ModalDialog {...propsLoading} />);

      expect((getByText('nope.avi') as HTMLButtonElement).disabled).toBe(true);
      expect((getByText('Do it.') as HTMLButtonElement).disabled).toBe(true);
      expect((getByText('−') as HTMLButtonElement).disabled).toBe(true);
    });
  });
});
