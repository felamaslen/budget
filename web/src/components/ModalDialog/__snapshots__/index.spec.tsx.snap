// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDialog /> should render a form list 1`] = `
<ul
  class="sc-kEYyzF hCTcAT"
  data-testid="form-list"
>
  <li
    class="sc-kkGfuU leTnSG"
  >
    <span
      class="sc-iAyFgw dKiNCX"
    >
      item
    </span>
    <div
      class="sc-frDJqD hdaOXn"
    >
      <input
        type="text"
        value="some item"
      />
    </div>
  </li>
  <li
    class="sc-kkGfuU leTnSG"
  >
    <span
      class="sc-iAyFgw dKiNCX"
    >
      cost
    </span>
    <div
      class="sc-frDJqD ggTVzp"
    >
      <input
        step="0.01"
        type="number"
        value="3.42"
      />
    </div>
  </li>
</ul>
`;
